File: 994_Rotting_Oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        
        minutes = 0
        numOfFreshOranges = 0
        
        queue = deque()
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 2:
                    queue.append([i, j])
                elif grid[i][j] == 1:
                    numOfFreshOranges += 1
                    
        directions = [ [-1, 0],
                       [0, 1],
                       [1, 0],
                       [0, -1]
                     ]
        
        while queue and numOfFreshOranges > 0:
            idx = len(queue)
                
            for _ in range(idx):
                curPos = queue.popleft()

                for direction in directions:
                    curRow = curPos[0] + direction[0]
                    curCol = curPos[1] + direction[1]

                    if curRow < 0 or curRow >= len(grid) or curCol < 0 or curCol >= len(grid[0]) or grid[curRow][curCol] == 0 or grid[curRow][curCol] == 2:
                        continue
                            
                    queue.append([curRow, curCol])
                    grid[curRow][curCol] = 2
                    numOfFreshOranges -= 1
                    
            minutes += 1
        
        return minutes if numOfFreshOranges == 0 else -1
        
        
            

File: swift/994_rotting_oranges.swift
class Solution {
  func orangesRotting(_ grid: [[Int]]) -> Int {
    var grid = grid
    let rowLen = grid.count
    let colLen = grid[0].count

    var rottenOrangesPos: [(row: Int, column: Int)] = []
    var minimumMinutes: Int = 0

    for row in 0..<rowLen {
      for column in 0..<colLen {
        if grid[row][column] == 2 {
          rottenOrangesPos.append((row, column))
        }
      }
    }

    bfs()

    return checkExistRottenOrange(grid) ? -1 : minimumMinutes

    func bfs() {
      var queue: [(row: Int, column: Int)] = []

      for rottenOrangePos in rottenOrangesPos {
        queue.append(rottenOrangePos)
      }

      let directions = [(-1, 0), (1, 0), (0, 1), (0, -1)]

      while !queue.isEmpty {
        let queueLen = queue.count

        for _ in 0..<queue.count {
          let curPos = queue.removeFirst()
          let curRow = curPos.row
          let curCol = curPos.column

          if grid[curRow][curCol] == 2 {
            for direction in directions {
              let newRow = curRow + direction.0
              let newCol = curCol + direction.1

              if 0 <= newRow && newRow < rowLen && 0 <= newCol && newCol < colLen
                && grid[newRow][newCol] == 1
              {
                grid[newRow][newCol] = 2
                queue.append((newRow, newCol))
              }
            }
          }
        }

        if !queue.isEmpty {
          minimumMinutes += 1
        }
      }

      return
    }

    func checkExistRottenOrange(_ grid: [[Int]]) -> Bool {
      for row in 0..<grid.count {
        for column in 0..<grid[0].count {
          if grid[row][column] == 1 {
            return true
          }
        }
      }

      return false
    }
  }
}
